### ISSUE TITLE
gem install rarely causes makeSharedShape() can only be called on non-shared shapes. (UnsupportedOperationException)

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1601]
When I tried to install `hamlit`, I got in TruffleRuby 1.0.0 RC13:

```
gem install hamlit
ERROR:  While executing gem ... (RuntimeError)
    makeSharedShape() can only be called on non-shared shapes. (UnsupportedOperationException)
	from com.oracle.truffle.object.ShapeImpl.makeSharedShape(ShapeImpl.java:956)
	from org.truffleruby.language.objects.shared.ShareObjectNode.createSharedShape(ShareObjectNode.java:116)
	from org.truffleruby.language.objects.shared.ShareObjectNodeGen.executeAndSpecialize(ShareObjectNodeGen.java:95)
	from org.truffleruby.language.objects.shared.ShareObjectNodeGen.executeShare(ShareObjectNodeGen.java:61)
	from org.truffleruby.language.objects.shared.WriteBarrierNode.writeBarrierCached(WriteBarrierNode.java:47)
	from org.truffleruby.language.objects.shared.WriteBarrierNodeGen.executeWriteBarrier(WriteBarrierNodeGen.java:46)
	from org.truffleruby.language.objects.shared.ShareInternalFieldsNode.shareCachedObjectArray(ShareInternalFieldsNode.java:50)
	from org.truffleruby.language.objects.shared.ShareInternalFieldsNodeGen.executeShare(ShareInternalFieldsNodeGen.java:50)
	from org.truffleruby.language.objects.shared.ShareObjectNode.shareCached(ShareObjectNode.java:62)
	from org.truffleruby.language.objects.shared.ShareObjectNodeGen.executeShare(ShareObjectNodeGen.java:43)
Translated to internal error
```

The same error when I tried something else in the morning, but I don't remember what's it.

**Comment by chrisseaton** [comment]
@eregon will know the most about this one.

**Comment by eregon** [comment]
I cannot reproduce, and instead get a compiler error:
```
compiling hamlit.c
/home/eregon/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/hamlit-2.9.3/ext/hamlit/hamlit.c:286:5: error: statement requires
      expression of integer type ('VALUE' (aka 'void *') invalid)
    switch (value) {
    ^       ~~~~~
/home/eregon/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/hamlit-2.9.3/ext/hamlit/hamlit.c:405:3: error: statement requires
      expression of integer type ('VALUE' (aka 'void *') invalid)
  switch (value) {
  ^       ~~~~~
2 errors generated.
make: *** [Makefile:214: hamlit.bc] Error 1
```

However, I think I can do something to fix the exception you got.

**Comment by deepj** [comment]
@eregon `gem install hamlit` is just a case when this was occurred. This issue have happened twice in different cases until now. But not in RC15 so far yet.

**Comment by eregon** [comment]
I merged a fix for this in aa6623d0c9aeb74f8465c414e5bd78d6501905bd, which was in the 19.0 release.
@deepj Thank you for the report.

**Comment by eregon** [comment]
Reopening, this needs a second fix.

**Comment by eregon** [comment]
This should be solved for good now with 22dfb295ee78d3f6de24253a835212dd21bb7d5b, it will be in the next release.
